src/openms/source/ANALYSIS/OPENSWATH/PeakIntegrator.cpp

```diff
@@ -375,4 +375,706 @@ namespace OpenMS
+    UInt info_iter_threshold { 1 };
+
+    // Learning rates (used in gradient descent and iRprop+)
+    double lr_h, lr_mu, lr_sigma, lr_tau;
+    lr_h = lr_mu = lr_sigma = lr_tau = 0.0125; // iRprop+ paper recommends 0.0125 // TODO: improve references to papers, maybe with something like [REF_NUM]
+
+    // Variables to limit the change in position `mu`
+    const double mu_max_dist = computeMuMaxDistance(TrX);
+    const double mu_left_boundary { mu - mu_max_dist };
+    const double mu_right_boundary { mu + mu_max_dist };
+
+    // The standard deviation between a selection of the precedent Es is computed.
+    // If said standard deviation is lower than a certain value,
+    // the computation of gradient descent is terminated
+    const Size last_few_Es_dim { 10 };
+    std::vector<double> last_few_Es(last_few_Es_dim, 0.0);
+    Size last_few_Es_idx = 0;
+    const double Es_std_dev_min = 1.0; // TODO: improve this value?
+
+    if (print_debug_ == 1)
+    {
+      std::cout << "GRADIENT DESCENT\nInput vectors size: " << xs.size() << "; Training set size: " << TrX.size() << std::endl;
+      std::cout << "The possible mu range is [" << mu_left_boundary << " " << mu_right_boundary << "]" << std::endl;
+    }
+
+    while (++iter_idx <= max_num_iter)
+    {
+      // Break if parameters are `nan` or `inf`
+      if (
+        std::isnan(h) || std::isnan(mu) || std::isnan(sigma) || std::isnan(tau) ||
+        std::isinf(h) || std::isinf(mu) || std::isinf(sigma) || std::isinf(tau)
+      )
+      {
+        std::cout << std::endl << "[" << iter_idx << "]" << std::endl;
+        std::cout << "One or more parameters are invalid." << std::endl;
+        std::cout << "Bad: h=" << h << " mu=" << mu << " sigma=" << sigma << " tau=" << tau << std::endl;
+        break;
+      }
+
+      // Compute the cost given the current parameters
+      const double current_E = Loss_function(TrX, TrY, h, mu, sigma, tau);
+
+      // Break if the computed cost is an invalid value
+      if (std::isnan(current_E) || std::isinf(current_E))
+      {
+        std::cout << std::endl << "[" << iter_idx << "]" << std::endl;
+        std::cout << "Bad: E value is invalid. current_E=" << current_E << std::endl;
+        break;
+      }
+
+      // If the current iteration is the best one, save the relevant values
+      if (current_E < best_E)
+      {
+        best_h = h;
+        best_mu = mu;
+        best_sigma = sigma;
+        best_tau = tau;
+        best_E = current_E;
+        best_iter = iter_idx;
+      }
+
+      // Compute the partial derivatives given the current parameters
+      double diff_E_h = E_wrt_h(TrX, TrY, h, mu, sigma, tau);
+      double diff_E_mu = E_wrt_mu(TrX, TrY, h, mu, sigma, tau);
+      double diff_E_sigma = E_wrt_sigma(TrX, TrY, h, mu, sigma, tau);
+      double diff_E_tau = E_wrt_tau(TrX, TrY, h, mu, sigma, tau);
+
+      // Logging info to the terminal
+      if (print_debug_ == 1 && iter_idx % info_iter_threshold == 0)
```

Comment:
May want to convert the std:cout to Log_debug

Reply:
I prefer to do this at the end, because I still haven't found a way to visualize `LOG_DEBUG` output in terminal

Reply:
Got it.